testlib/s3mock.py

```diff
@@ -307,60 +229,68 @@ class S3Mock:
     After all the steps and other things that you want to do are done, then you
     can assert that the entire expected conversation has occurred.
 
-    **Recording**
 
-    One of the difficulties with writing tests that assert HTTP conversations
-    happen in a certain way is that if the conversation happens over SSL, then
-    it's not readable using network sniffing tools. Thus ``S3Mock`` provides a
-    record feature that spits out what's going on to the specified file or
-    ``s3mock.log``.
+    **Troubleshooting**
+
+    If ``S3Mock`` is thwarting you, you can tell it to ``run_on_error`` and it'll
+    execute the urlopen and tell you want the response was. It'll also tell you
+    what it expected. This helps debugging assertions on HTTP conversations.
 
     Usage::
 
         with S3Mock() as s3:
-            s3.record(filename='s3mock.log')
-
-            # Do stuff here
+            s3.run_on_error()
+            # ... add steps, etc
 
     """
 
     def __init__(self):
-        self.adapter = FakeAdapter()
+        # The expected conversation specified by the developer
+        self.expected_conv = []
 
-    def __enter__(self):
-        self.start_mock()
-        return self
+        # The actual conversation that happened
+        self.conv = []
 
-    def __exit__(self, exc_type, exc_value, traceback):
-        self.stop_mock()
+        self._patcher = None
 
-    def _get_recording_adapter(self, session, url):
-        recording_adapter = RecordingAdapterShim()
-        recording_adapter.wrapped_adapter = self._real_get_adapter(
-            self=session, url=url
-        )
-        recording_adapter.filename = self._filename_to_record
-        return recording_adapter
+        self._run_on_error = False
 
-    def record(self, filename="s3mock.log"):
-        """Starts an HTTP conversation recording session
+    def mocked_urlopen(self, pool, method, url, body=None, headers=None, **kwargs):
+        req = Request(method, url, body, headers, pool.scheme, pool.host, pool.port)
 
-        :arg str filename: The name of the file to log to.
+        if self.expected_conv and self.expected_conv[0].match(req):
+            step = self.expected_conv.pop(0)
+            resp = step.build_response(req)
+            self.conv.append((req, step, resp))
+            return resp
 
-        """
-        if os.path.isfile(filename):
-            os.remove(filename)
+        # NOTE(willkg): We use print here because fiddling with the logging
+        # framework inside test scaffolding is "tricky".
+        print("THWARTED SEND:\nHTTP Request:\n%s" % serialize_request(req))
+        if self.expected_conv:
+            step = self.expected_conv[0]
+            print("Expected:\n%s %s\n%s" % (step.method, step.url, step.body))
+        else:
+            print("Expected: nothing")
+
+        if self._patcher and self._run_on_error:
+            resp = self._patcher.get_original()[0](
```

Comment:
This calls the original `urlopen` on the pool instance which will perform the HTTP request and tell you what you got.